### DISCUSSION TITLE
Signaling a page for SSR / SSG

### DISCUSSION THREAD

**Discussion by matthewp** [withastro/roadmap#133] [Proposal]
# Background 

I've been working on SSR support for Astro and have made a low-level (and likely buggy) preview available with the `--experimental-ssr` flag. 

As part of this work I've done research on how various frameworks signal that a page should be SSRed or statically rendered (rendered at build time). The research is in [this spreadsheet](https://docs.google.com/spreadsheets/d/1Rr_X0r0d0sn8ig3kBJfKCCdyckDQq30E3Dm4jiZoxIU/edit#gid=0).

# Recommendation

In summary I am recommending that:

* The decision to SSR or not is based on which __adapter__ you use, with a static adapter (SSG) being the default.
* If you use an adapter that supports SSR, such as a Cloudflare adapter, the build will be to SSR mode.
* An opt-in `export default staticRender = true` could be added in the future to statically render a specific page.

# Explanation

Given the nature of Astro and in particular the fact that:

* Astro components only run on the server.
* Astro components support top-level-await.

I think we should preserve the *flat* nature of Astro. Meaning that you can do this:

```astro
---
let pets = await fetch('/api/pets');
---

{pets.map(pet => (<span> { pet.name }</span>))}
```

Given that, adopting an exported function to signal SSR, whether that be `getServerSideProps()` or `load()` or `setup()` seems unnecessary and potentially confusing.

## Adapters

We'll develop an adapter API for this new support. A SSG adapter will be built using this API but statically render.

It's unknown what this API will look like, but the adapter would have the ability to render the pages. This means that even host adapters *could* also support static rendering as well.

## getStaticPaths

The conclusion to this would mean that `getStaticPaths` is *not needed* for apps that build to SSR. Instead write flat Astro as you would in any non-dynamic page.

getStaticPaths remains useful when building to static; it still makes sense for that use-case.

What's also nice here is that this change could be completely non-breaking and therefore not hold up the 1.0 release.

# Drawbacks / Unknowns

Some SSR features *might* not work well with a flat design. These include:

- Redirects
- Setting a static code

We don't know yet and need to prototype some ideas to see how they look/feel in a flat structure.

**Comment by ghost** [comment]
I like everything mentioned here! I especially like the concept of "adapters." This ensures builds are static out-of-the-box, with an opt-in for SSR functionality in your deployment target of choice.

However, I'm getting a _bit_ concerned with the number of magic imports we've been introducing. Instead of introducing a `export const staticRender = true`, maybe we could consider a standardized `export const meta` for these use cases? It would also let us maintain a single TypeScript type as more meta attributes are introduced. For example:

```ts
import type { Meta } from 'astro'

export const meta: Meta = {
  render: {
    mode: 'ssr',
    // ex. allow cache headers
    headers: {...}
  }
}
```
☝️ This could behave similarly to frontmatter in your markdown, and be statically analyzable on SSR routes. Likely worth a separate RFC if we pursue this though.
